calico_kubernetes/calico_kubernetes.py

```diff
@@ -359,16 +410,72 @@ def _apply_tags(self, profile_name, pod):
             print('Error: Could not apply tags. Profile %s could not be found. Exiting' % profile_name)
             sys.exit(1)
 
-        for k, v in labels.iteritems():
-            tag = '%s_%s' % (k, v)
-            tag = tag.replace('/', '_')
-            print('Adding tag ' + tag)
-            profile.tags.add(tag)
-
+        # Grab namespace and create a tag if it exists.
+        namespace, ns_tag = self._get_namespace_and_tag(pod)
+
+        if ns_tag:
+            print('Adding tag %s' % ns_tag) 
+            profile.tags.add(ns_tag)
+           
+        # Create tags from labels
+        labels = self._get_metadata(pod, 'labels')
+        if labels:
+            for k, v in labels.iteritems():
+                tag = self._label_to_tag(k, v, namespace)
+                print('Adding tag ' + tag)
+                profile.tags.add(tag)
+        else: 
+            print('No labels found in pod %s' % pod)
+               
         self._datastore_client.profile_update_tags(profile)
 
         print('Finished applying tags.')
 
+    def _get_metadata(self, pod, key):
+        """
+        Return Metadata[key] Object given Pod
+        Returns None if no key-value exists
+        """
+        try:
+            val = pod['metadata'][key]
+        except KeyError:
+            print('No %s found in pod %s' % (key, pod))
+            return None
+
+        print("%s of pod %s:\n%s" % (key, pod, val))
+        return val
+
+    def _escape_chars(self, unescaped_string):
+        """
+        Calico can only handle 3 special chars, '_.-'
+        This function uses regex sub to replace SCs with '_'
+        """
+        escape_seq = '_'
+        return re.sub('[^a-zA-Z0-9\.-]', escape_seq, unescaped_string)
+
+    def _get_namespace_and_tag(self, pod):
+        namespace = self._get_metadata(pod, 'namespace')
+        ns_tag = self._escape_chars('%s=%s' % ('namespace', namespace)) if namespace else None
+        return namespace, ns_tag
+
+    def _label_to_tag(self, label_key, label_value, namespace):
+        """
+        Labels are key-value pairs, tags are single strings. This function handles that translation
+        1) concatenate key and value with '='
+        2) prepend a pod's namespace followed by '/' if available
+        3) replace special characters with urllib-style escape sequence
```

Comment:
Third bullet point here is wrong - can just say "escape the generated string so it is Calico compatible" -- the implementation of the escape logic is irrelevant.